File: ruby/examples/xrates.rb
#!/usr/bin/env ruby

this_dir = File.expand_path(File.dirname(__FILE__))
lib_dir = File.dirname(this_dir)
$LOAD_PATH.unshift(lib_dir) unless $LOAD_PATH.include?(lib_dir)

require 'grpc'
require 'xrates_services'

include Xrates

def main
  stub = XRates::Stub.new('localhost:50800')

  puts "Getting one currency"
  ['MYR', 'UNK'].each do |c|
    begin
      rate = stub.get(Currency.new(currency: c))
      puts "Rate for MYR: #{rate.rate}"
    rescue GRPC::BadStatus => b
      puts "Couldn't get currency. #{b}"
    end
  end

  puts "Getting multiple currencies"
  rates = stub.all(Currencies.new(currencies: []))
  rates.rates.each do |cur, rate|
    puts "Rate for #{cur}: #{rate}"
  end
end

main

File: examples/xrates.go
package main

import (
	"flag"
	"log"
	"zumata/xrates"

	"golang.org/x/net/context"
	"google.golang.org/grpc"
)

func main() {
	server := flag.String("server", "", "Server which host the service")
	flag.Parse()
	if *server == "" {
		log.Fatal("Server cannot be empty.")
	}

	var opts []grpc.DialOption
	conn, err := grpc.Dial(*server, opts...)
	if err != nil {
		log.Fatalf("Fail to dial: %v", err)
	}
	defer conn.Close()

	client := xrates.NewXRatesClient(conn)

	log.Println("Getting one currency")
	for _, c := range []string{"MYR", "UNK"} {
		rate, err := client.Get(context.Background(), &xrates.Currency{c})
		if err != nil {
			log.Printf("Couldn't get currency. %v\n", err)
		} else {
			log.Printf("Rate for MYR: %f\n", rate.Rate)
		}
	}

	log.Println("Getting multiple currencies")
	rates, err := client.All(context.Background(), &xrates.Currencies{[]string{}})
	if err != nil {
		log.Printf("Couldn't get currency. %v\n", err)
	} else {
		for cur, rate := range rates.Rates {
			log.Printf("Rate for %s: %f\n", cur, rate)
		}
	}
}
